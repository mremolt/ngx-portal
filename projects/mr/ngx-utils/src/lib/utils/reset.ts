import { Action, ActionReducer } from '@ngrx/store';

export const APP_RESET = '[Application] Reset State';

export class AppReset implements Action {
  public readonly type = APP_RESET;
}

export function resetMetaReducer(reducer: ActionReducer<any>): ActionReducer<any> {
  return function(state, action) {
    if (action.type === APP_RESET) {
      return reducer(undefined, action);
    }

    return reducer(state, action);
  };
}
