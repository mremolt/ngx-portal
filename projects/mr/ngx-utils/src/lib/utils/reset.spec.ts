import { Action, ActionReducer } from '@ngrx/store';

import { AppReset, resetMetaReducer } from './reset';

describe('App reset', () => {
  describe('AppReset', () => {
    it('builds the correct action', () => {
      const action = new AppReset();
      expect(action).toMatchSnapshot();
    });
  });

  describe('resetMetaReducer', () => {
    let reducer: ActionReducer<any>;
    let mockReducer: any;
    const anyAction: Action = { type: 'TEST' };
    const anyState: any = { number: 42 };

    beforeEach(() => {
      mockReducer = jest.fn().mockImplementation((state: any, _action: any) => state);
      reducer = resetMetaReducer(mockReducer);
    });

    it('calls the given reducer with undefined state', () => {
      const newState = reducer(anyState, new AppReset());
      expect(mockReducer).toHaveBeenCalledWith(undefined, expect.any(AppReset));
      expect(newState).toBeUndefined();
    });

    it('calls the given reducer with given state', () => {
      const newState = reducer(anyState, anyAction);
      expect(mockReducer).toHaveBeenCalledWith(anyState, anyAction);
      expect(newState).toBe(anyState);
    });
  });
});
