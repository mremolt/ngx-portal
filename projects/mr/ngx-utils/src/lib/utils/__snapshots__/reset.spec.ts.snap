// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App reset AppReset builds the correct action 1`] = `
AppReset {
  "type": "[Application] Reset State",
}
`;
